cmd/geoipupdate/main.go

```diff
@@ -0,0 +1,428 @@
+
+	if verbose {
+		log.Printf("Performing get filename request to %s", url)
+	}
+	res, err := client.Get(url)
+	if err != nil {
+		return "", errors.Wrap(err, "error performing HTTP request")
+	}
+	defer func() {
+		if err := res.Body.Close(); err != nil {
+			log.Fatalf("Error closing response body: %+v", errors.Wrap(err, "closing body"))
+		}
+	}()
+
+	buf, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		return "", errors.Wrap(err, "error reading response body")
+	}
+
+	if res.StatusCode != http.StatusOK {
+		return "", errors.Errorf("unexpected HTTP status code: %s: %s", res.Status, buf)
+	}
+
+	if len(buf) == 0 {
+		return "", errors.New("response body is empty")
+	}
+
+	if bytes.Count(buf, []byte("\n")) > 0 ||
+		bytes.Count(buf, []byte("\x00")) > 0 {
+		return "", errors.New("invalid characters in filename")
+	}
+
+	return string(buf), nil
+}
+
+const zeroMD5 = "00000000000000000000000000000000"
+
+func getCurrentMD5(
+	config *Config,
+	verbose bool,
+	filename string,
+) (string, error) {
+	path := filepath.Join(config.DatabaseDirectory, filename)
+
+	fh, err := os.Open(path)
+	if err != nil {
+		if os.IsNotExist(err) {
+			if verbose {
+				log.Printf("Not calculating MD5 sum as file does not exist: %s", path)
+			}
+			return zeroMD5, nil
+		}
+		return "", errors.Wrap(err, "error opening file")
+	}
+	defer func() {
+		if err := fh.Close(); err != nil {
+			log.Fatalf("Error closing file: %+v", errors.Wrap(err, "closing file"))
+		}
+	}()
+
+	fi, err := fh.Stat()
+	if err != nil {
+		return "", errors.Wrap(err, "error stat'ing file")
+	}
+	if !fi.Mode().IsRegular() {
+		return "", errors.New("not a regular file")
+	}
+
+	h := md5.New()
+	if _, err := io.Copy(h, fh); err != nil {
+		return "", errors.Wrap(err, "error reading file")
+	}
+	sum := fmt.Sprintf("%x", h.Sum(nil))
+	if verbose {
+		log.Printf("Calculated MD5 sum for %s: %s", path, sum)
+	}
+	return sum, nil
+}
+
+func maybeUpdate(
+	config *Config,
+	verbose bool,
+	editionID,
+	filename,
+	md5 string,
+) error {
+	url := fmt.Sprintf(
+		"%s/geoip/databases/%s/update?db_md5=%s",
+		config.URL,
+		url.PathEscape(editionID),
+		url.QueryEscape(md5),
+	)
+
+	req, err := http.NewRequest(http.MethodGet, url, nil)
+	if err != nil {
+		return errors.Wrap(err, "error creating request")
+	}
+	if !strings.HasPrefix(editionID, "GeoLite2") {
+		req.SetBasicAuth(fmt.Sprintf("%d", config.AccountID), config.LicenseKey)
+	}
+
+	if verbose {
+		log.Printf("Performing update request to %s", url)
+	}
+	res, err := client.Do(req)
+	if err != nil {
+		return errors.Wrap(err, "error performing HTTP request")
+	}
+	defer func() {
+		if err := res.Body.Close(); err != nil {
+			log.Fatalf("Error closing response body: %+v", errors.Wrap(err, "closing body"))
+		}
+	}()
+
+	if res.StatusCode == http.StatusNotModified {
+		if verbose {
+			log.Printf("No new updates available for %s", editionID)
+		}
+		return nil
+	}
+
+	if res.StatusCode != http.StatusOK {
+		buf, err := ioutil.ReadAll(res.Body)
```

Comment:
We might want to limit the size on this.

Reply:
Yep, good idea!